### PR TITLE
Fix the policy of foreign tables

### PR DESCRIPTION
**Pull request by adam8157**
In some special FDW cases, Greenplum could have more or less backends than the number of actual segments in one gang, this commit sets the policy correctly.

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [x] Pass `make installcheck`
- [x] Review a PR in return to support the community

### CODE REVIEW

**Review by jingwen-yang-yjw — COMMENTED** [review]
We could add a test case that the value of option num_segments is larger than the number of segments of local gpdb here.

File: contrib/postgres_fdw/expected/gp_postgres_fdw_optimizer.out [hunk 1]
```diff
@@ -131,21 +131,22 @@ alter server loopback options(add num_segments '2');
...
                             QUERY PLAN                             
 -------------------------------------------------------------------
- Gather Motion 3:1  (slice1; segments: 3)
+ Gather Motion 2:1  (slice1; segments: 2)
    Output: f1, f2, f3
    ->  Foreign Scan on postgres_fdw_gp.gp_ft1
          Output: f1, f2, f3
          Remote SQL: SELECT f1, f2, f3 FROM postgres_fdw_gp."GP 1"
- Optimizer: Pivotal Optimizer (GPORCA)
+ Optimizer: Postgres query optimizer
```
Comment (avamingli) [inline contrib/postgres_fdw/expected/gp_postgres_fdw_optimizer.out:139]: ORCA fallback?

Comment (adam8157) [reply contrib/postgres_fdw/expected/gp_postgres_fdw_optimizer.out]: Correct, ORCA doesn't support gangs with more or less QEs than the segments count.

File: src/backend/commands/explain.c [hunk 2]
```diff
@@ -1805,8 +1805,8 @@ ExplainNode(PlanState *planstate, List *ancestors,
 
 				Assert(plan->lefttree);
 
-				motion_snd = list_length(es->currentSlice->segments);
-				motion_recv = parentSlice == NULL ? 1 : list_length(parentSlice->segments);
+				motion_snd = es->currentSlice->planNumSegments;
+				motion_recv = parentSlice == NULL ? 1 : parentSlice->planNumSegments;
```
Comment (zhaorui20005) [inline src/backend/commands/explain.c:1809]: It seems that only when direct dispatch is triggered, then it makes a difference. Is it possible to add one test case to that?

File: src/backend/optimizer/plan/createplan.c [hunk 3]
```diff
@@ -3100,10 +3100,9 @@ create_motion_plan(PlannerInfo *root, CdbMotionPath *path)
 			break;
 
 		case CdbLocusType_General:
-			/*  */
 			sendSlice->gangType = GANGTYPE_SINGLETON_READER;
 			sendSlice->numsegments = 1;
-			sendSlice->segindex = gp_session_id % getgpsegmentCount();
+			sendSlice->segindex = gp_session_id % subpath->locus.numsegments;
```
Comment (zhaorui20005) [inline src/backend/optimizer/plan/createplan.c:3105]: segindex has a range from 0 to getgpsegmentCount()-1, if % subpath->locus.numsegments, will segindex be larger than `getgpsegmentCount()-1`?

File: src/backend/commands/explain.c (hunk 2 above)
Comment (adam8157) [reply src/backend/commands/explain.c]: You are right, this change is removed.

File: src/backend/optimizer/plan/createplan.c (hunk 3 above)
Comment (adam8157) [reply src/backend/optimizer/plan/createplan.c]: General policy's `numsegments` is -1, this change is a part of my other refactor and included by mistake, removed now.

**Comment by zhaorui20005** [comment]
All of this current codes of this PR looks good to me. But it needs more reviewers. 
And does this PR need to consider to open two-phase transaction since it involves inserting into Foreign table on different segments?

**Comment by adam8157** [comment]
> And does this PR need to consider to open two-phase transaction since it involves inserting into Foreign table on different segments?

That deserves another PR for a separate discussion.

**Review by zhaorui20005 — APPROVED** [review]
LGTM, The only risk point of this PR is the modification for nodeModifyTable.c.
I have tested that when num_segment is larger than 3, the writing plan and result is correct. And I have opened debug and wal logs to prove that altering foreign table does not writes XLOG.

File: src/backend/executor/nodeModifyTable.c [hunk 4]
```diff
@@ -2375,7 +2375,21 @@ ExecModifyTable(PlanState *pstate)
...
+	segid_attno = resultRelInfo->ri_segid_attno;
+
+	/*
+	 * Check cdbcomponent_allocateIdleQE()
+	 *
+	 * In some cases Greenplum has more than one QE on on segment writing the
+	 * external data via foreign tables.
+	 */
+	if (Gp_role == GP_ROLE_EXECUTE && !Gp_is_writer &&
+		!resultRelInfo->ri_RelationDesc->rd_rel->relkind == RELKIND_FOREIGN_TABLE)
```
Comment (kainwen) [inline src/backend/executor/nodeModifyTable.c:2392]: Do you want to mean:

```c
resultRelInfo->ri_RelationDesc->rd_rel->relkind != RELKIND_FOREIGN_TABLE
```

here?

Comment (adam8157) [reply src/backend/executor/nodeModifyTable.c]: 😂

File: src/backend/executor/nodeModifyTable.c [hunk 5]
```diff
@@ -2375,7 +2375,21 @@ ExecModifyTable(PlanState *pstate)
...
+	segid_attno = resultRelInfo->ri_segid_attno;
+
+	/*
+	 * Check cdbcomponent_allocateIdleQE()
+	 *
+	 * In some cases Greenplum has more than one QE on on segment writing the
+	 * external data via foreign tables.
+	 */
+	if (Gp_role == GP_ROLE_EXECUTE && !Gp_is_writer &&
+		resultRelInfo->ri_RelationDesc->rd_rel->relkind != RELKIND_FOREIGN_TABLE)
```
Comment (kainwen) [inline src/backend/executor/nodeModifyTable.c:2392]: @zhaorui20005 per offline discussion, you mention you try that Modify to Foreign table , GPDB side (QE) will not generate XLOG or genreate local XID.

Can you try to post what you find here and read the code to verify again? Thanks.

**Review by kainwen — APPROVED** [review]
Leave one comment, others LGTM.

File: src/backend/executor/nodeModifyTable.c (hunk 5 above)
Comment (zhaorui20005) [reply src/backend/executor/nodeModifyTable.c:2392]: I made one comparison between heap tuple and FDW table with this branch (https://github.com/zhaorui20005/gpdb/tree/enable_fdw_table_insert_and_2phase), and this branch not only enables the modification for postgres_fdw Foreign table insert with multi_hosts and multi_ports, but also enables the two-phase commit with this commit(https://github.com/zhaorui20005/gpdb/commit/172120023fa159869275b0c8def8411d070d5038).
Here is the sql to create the foreign table.

**Pull request merged**